### PR TITLE
Extensible constraint interface

### PR DESCRIPTION
**Pull request by PhilipVinc**
This PR greatly simplifies the usage of custom constraints with our Hilbert spaces, and solves some usability issues with them.
See the example:

```python
import jax
import jax.numpy as jnp

import netket as nk
from netket.utils import struct

class SumConstraint(nk.hilbert.constraint.DiscreteHilbertConstraint):
    # A simple constraint checking that the total sum of the elements
    # in the configuration is equal to a given value.

    total_sum : float = struct.field(pytree_node=False)

    def __init__(self, total_sum):
        self.total_sum = total_sum

    def __call__(self, x):
        # Makes it jax-compatible
       return jnp.sum(x, axis=-1) == self.total_sum

    def __hash__(self):
        return hash(("SumConstraint", self.total_sum))

    def __eq__(self, other):
        if isinstance(other, SumConstraint):
            return self.total_sum == other.total_sum
        return False

hi = nk.hilbert.Fock(n_max=2, N=5, constraint=SumConstraint(3))
hi.random_state(jax.random.key(3), 10)
```

Admittedly, the syntax to define a custom constraint is not terse. In the future we might be able to autogenerate the hash and eq methods, but for the moment I would keep it as this. 

A lot of examples and informative errors with examples are added everywhere, so it should be clear how to define those constraints. 

The old constraints will not be supported anymore. They were already unsupported, and did not work with random state generation...

 - A big add of this PR is a method to automatically generate random state within the constrained space. Previously we were always asking users to do it. Instead, in this PR there is an automatic fallback that generates random states from the unconstrained space until he gets one that satisfies the constraint. While inefficient, this is not a problem because in general we just need it to seed the sampler. To be safe, we throw a warning telling users to define their own random state logic because this one might be slow.

 - I also added a special constraint to merge two constraints. This can be used to combine multiple constraints.

The PR also refactors a bit the internal methods defined by @inailuig to make it more accessible to external users. 

I also deprecated `CustomHilbert`. It was not really playing out nicely with everything else we have, and it's time we kill it. There's not really any need for it anymore now. Fock + custom constraints should be enough for every use case.

### CODE REVIEW

**Review by gcarleo — APPROVED** [review]
nice!

**Comment by gcarleo** [comment]
I just wonder if we should put a time limit on the random generation. the default method will be exponentially slow in general for an arbitrary constraint, the risk is that the user does not know why the code is running forever ?

**Comment by gcarleo** [comment]
in my C++ code I had a limit on the number of maximum random trials, if the constraint wasn't satisfied it would error

**Comment by PhilipVinc** [comment]
Yeah, but how many? And this would be really hardcoded...

**Comment by gcarleo** [comment]
yes it's super ugly and hard coded ( could be an option that can be overriden though) I don't know... I'd put a maximum of 1000 trials haha. But ok, it's not crucial for this PR maybe, it's just to avoid some headaches to people who don't know it

**Comment by PhilipVinc** [comment]
It prints a very comprehensible warning. 
Assuming people know how to read, I don't think they can be found off guard, even if it ends up diverging.

```python
   ...: hi = nk.hilbert.Fock(n_max=2, N=5, constraint=SumConstraint(3))
   ...: hi.random_state(jax.random.key(3), 10)
/Users/filippo.vicentini/Dropbox/Ricerca/Codes/Python/netket/netket/hilbert/random/homogeneous.py:79: UnoptimisedCustomConstraintRandomStateMethodWarning:
Defaulting to a slow, possibly infinitely-looping method to generate random state of
the current Hilbert space with a custom constraint. Consider implementing a
custom `random_state` method for your constraint if this method takes a long time to
generate a random state.

To generate a custom random_state dispatched method, you should use multiple dispatch
following the following syntax:

>>> import netket as nk
>>> from netket.utils import dispatch
>>>
>>> @dispatch.dispatch
>>> def random_state(hilb: netket.hilbert.fock.Fock,
                    constraint: __main__.SumConstraint,
                    key,
                    batches: int,
                    *,
                    dtype=None):
>>>    # your custom implementation here
>>>    # You should return a batch of `batches` random states, with the given dtype.
>>>    # return jax.Array with shape (batches, hilb.size) and dtype dtype.
-------------------------------------------------------
For more detailed informations, visit the following link:
	 https://netket.readthedocs.io/en/latest/api/_generated/errors/netket.errors.UnoptimisedCustomConstraintRandomStateMethodWarning.html
or the list of all common errors and warnings at
	 https://netket.readthedocs.io/en/latest/api/errors.html
-------------------------------------------------------
```
